File: src/main/java/com/heekng/celloct/controller/ManageJoinRequestController.java
package com.heekng.celloct.controller;

import com.heekng.celloct.config.oauth.dto.SessionMember;
import com.heekng.celloct.dto.JoinRequestDto;
import com.heekng.celloct.dto.ShopDto;
import com.heekng.celloct.entity.Manager;
import com.heekng.celloct.entity.Shop;
import com.heekng.celloct.repository.JoinRequestRepository;
import com.heekng.celloct.repository.ManagerRepository;
import com.heekng.celloct.repository.ShopRepository;
import com.heekng.celloct.service.JoinRequestService;
import com.heekng.celloct.service.ShopService;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpSession;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Controller
@Slf4j
@RequiredArgsConstructor
@RequestMapping("/manage")
public class ManageJoinRequestController {

    private final HttpSession httpSession;
    private final ManagerRepository managerRepository;
    private final JoinRequestRepository joinRequestRepository;
    private final JoinRequestService joinRequestService;
    private final ShopService shopService;

    @GetMapping("/{shopId}/joinRequest")
    public String joinRequest(@PathVariable("shopId") Long shopId, Model model) {
        SessionMember member = (SessionMember) httpSession.getAttribute("member");
        Optional<Manager> managerOptional = managerRepository.findByMemberIdAndShopId(member.getId(), shopId);
        if (managerOptional.isEmpty()) {
            return "redirect:/";
        }

        List<JoinRequestDto.ManagerShopJoinRequestResponse> joinRequestResponses = joinRequestRepository.findWithMemberByShopId(shopId)
                .stream().map(JoinRequestDto.ManagerShopJoinRequestResponse::new)
                .collect(Collectors.toList());
        model.addAttribute("joinRequests", joinRequestResponses);
        Shop shop = shopService.findShop(shopId);
        model.addAttribute("shop", new ShopDto.ShopDetailResponse(shop));

        return "manager/manageShopJoinRequest";
    }

    @PostMapping("/{shopId}/joinRequest/approval")
    @ResponseBody
    public Boolean approval(@RequestBody JoinRequestDto.ApprovalRefusalRequest approvalRequest, @PathVariable("shopId") Long shopId) {
        SessionMember member = (SessionMember) httpSession.getAttribute("member");
        Optional<Manager> managerOptional = managerRepository.findByMemberIdAndShopId(member.getId(), shopId);
        if (managerOptional.isEmpty()) {
            return false;
        }

        Long staffId = joinRequestService.approval(approvalRequest);

        return true;
    }

    @PostMapping("/{shopId}/joinRequest/refusal")
    @ResponseBody
    public Boolean refusal(@RequestBody JoinRequestDto.ApprovalRefusalRequest approvalRequest, @PathVariable("shopId") Long shopId) {
        SessionMember member = (SessionMember) httpSession.getAttribute("member");
        Optional<Manager> managerOptional = managerRepository.findByMemberIdAndShopId(member.getId(), shopId);
        if (managerOptional.isEmpty()) {
            return false;
        }

        joinRequestService.refusal(approvalRequest);

        return true;
    }
}

File: src/main/kotlin/com/heekng/celloct/controller/ManageJoinRequestController.kt
package com.heekng.celloct.controller

import com.heekng.celloct.annotation.RoleCheck
import com.heekng.celloct.annotation.enums.UserType
import com.heekng.celloct.dto.JoinRequestDto
import com.heekng.celloct.dto.ShopDto
import com.heekng.celloct.repository.JoinRequestRepository
import com.heekng.celloct.repository.ManagerRepository
import com.heekng.celloct.repository.ShopRepository
import com.heekng.celloct.service.JoinRequestService
import com.heekng.celloct.util.findByIdOrThrow
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.web.bind.annotation.*
import javax.servlet.http.HttpSession

@Controller
@RequestMapping("/manage")
class ManageJoinRequestController(
    private val httpSession: HttpSession,
    private val managerRepository: ManagerRepository,
    private val joinRequestRepository: JoinRequestRepository,
    private val joinRequestService: JoinRequestService,
    private val shopRepository: ShopRepository,
) {


    @GetMapping("/{shopId}/joinRequest")
    @RoleCheck(UserType.MANAGER)
    fun joinRequest(
        @PathVariable("shopId") shopId: Long,
        model: Model,
    ): String {
        val joinRequestResponses = joinRequestRepository.findWithMemberByShopId(shopId)
            .map { joinRequest -> JoinRequestDto.ManagerShopJoinRequestResponse(joinRequest) }
        model.addAttribute("joinRequests", joinRequestResponses)
        val shop = shopRepository.findByIdOrThrow(shopId)
        model.addAttribute("shop", ShopDto.ShopDetailResponse(shop))

        return "manager/shopJoinRequest"
    }

    @PostMapping("/{shopId}/joinRequest/approval")
    @ResponseBody
    @RoleCheck(UserType.MANAGER, true)
    fun approval(
        @PathVariable("shopId") shopId: Long,
        @RequestBody approvalRequest: JoinRequestDto.ApprovalRefusalRequest,
    ): Boolean {
        joinRequestService.approval(approvalRequest)
        return true
    }

    @PostMapping("/{shopId}/joinRequest/refusal")
    @ResponseBody
    @RoleCheck(UserType.MANAGER, true)
    fun refusal(
        @PathVariable("shopId") shopId: Long,
        @RequestBody approvalRequest: JoinRequestDto.ApprovalRefusalRequest,
    ): Boolean {
        joinRequestService.refusal(approvalRequest)
        return true
    }



}